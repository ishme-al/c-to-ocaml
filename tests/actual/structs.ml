[@@@ocaml.warning "-26"]
[@@@ocaml.warning "-27"]
[@@@ocaml.warning "-32"]
[@@@ocaml.warning "-69"]

open Core

let rec set_at_index (lst : 'a list) (index : int) (value : 'a) : 'a list =
  match lst with
  | [] -> failwith "Index out of bounds"
  | hd :: tl ->
      if index = 0 then value :: tl else hd :: set_at_index tl (index - 1) value

type myStructure = { a : int; b : int; c : char; d : float }

let transformAB (str : myStructure) : int =
  let c : int = Int.( * ) str.a str.b in
  let b : int = Int.( + ) str.a str.b in
  Int.( + ) b c

let addAb (str : myStructure) : int = Int.( + ) str.a str.b
let multAb (str : myStructure) : int = Int.( * ) str.a str.b
let subAb (str : myStructure) : int = Int.( - ) str.a str.b
let divideAb (str : myStructure) : int = Int.( / ) str.a str.b

let () =
  let str : myStructure = { a = 1; b = 2; c = 'a'; d = 3. } in
  let a : int = transformAB str in
  let b : int = addAb str in
  let c : int = multAb str in
  let d : int = subAb str in
  let e : int = divideAb str in
  printf "a: %d\n" a;
  printf "b: %d\n" b;
  printf "c: %d\n" c;
  printf "d: %d\n" d;
  printf "e: %d\n" e;
  exit 0
