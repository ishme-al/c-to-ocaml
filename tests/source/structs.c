#include <stdio.h>

struct myStructure {
    int a;
    int b;
    char c;
    float d;
};

int transformAB ( struct myStructure str) {
    int c = str.a * str.b;
    int b = str.a + str.b;
    // int c = str.a * str.b;
    return b + c;
};

int addAb ( struct myStructure str) {
    return str.a + str.b;
};

int multAb ( struct myStructure str) {
    return str.a * str.b;
};

int subAb ( struct myStructure str) {
    return str.a - str.b;
};

int divideAb ( struct myStructure str) {
    return str.a / str.b;
};


int main() {
    struct myStructure str = {1, 2, 'a', 3.0};
    int a = transformAB(str);
    int b = addAb(str);
    int c = multAb(str);
    int d = subAb(str);
    int e = divideAb(str);
    printf("a: %d\n", a);
    printf("b: %d\n", b);
    printf("c: %d\n", c);
    printf("d: %d\n", d);
    printf("e: %d\n", e);
    return 0;
}