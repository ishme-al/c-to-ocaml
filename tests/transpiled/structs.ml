type myStructure = { a : int; b : int; c : char; d : float }

let transformAB (str : myStructure) : int =
  let c : int = str.a * str.b in
  let b : int = str.a + str.b in
  b + c

let addAb (str : myStructure) : int = str.a + str.b
let multAb (str : myStructure) : int = str.a * str.b
let subAb (str : myStructure) : int = str.a - str.b
let divideAb (str : myStructure) : int = str.a / str.b
