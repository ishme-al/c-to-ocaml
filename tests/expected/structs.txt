a: 5
b: 3
c: 2
d: -1
e: 0
